libopeniscsiusr/idbm.c

```diff
@@ -0,0 +1,699 @@
+		   IDBM_SHOW, num, _CAN_MODIFY);
+	_rec_str(IFACE_TCP_TIMESTAMP, recs, iface, tcp_timestamp, IDBM_SHOW,
+		 num, _CAN_MODIFY);
+	_rec_str(IFACE_REDIRECT, recs, iface, redirect, IDBM_SHOW, num,
+		 _CAN_MODIFY);
+	_rec_uint16(IFACE_DEF_TMF_TMO, recs, iface, def_task_mgmt_tmo,
+		    IDBM_SHOW, num, _CAN_MODIFY);
+	_rec_str(IFACE_HDRDGST, recs, iface, header_digest, IDBM_SHOW, num,
+		 _CAN_MODIFY);
+	_rec_str(IFACE_DATADGST, recs, iface, data_digest, IDBM_SHOW, num,
+		 _CAN_MODIFY);
+	_rec_str(IFACE_IMM_DATA, recs, iface, immediate_data, IDBM_SHOW, num,
+		 _CAN_MODIFY);
+	_rec_str(IFACE_INITIAL_R2T, recs, iface, initial_r2t, IDBM_SHOW, num,
+		 _CAN_MODIFY);
+	_rec_str(IFACE_DSEQ_INORDER, recs, iface, data_seq_inorder, IDBM_SHOW,
+		 num, _CAN_MODIFY);
+	_rec_str(IFACE_DPDU_INORDER, recs, iface, data_pdu_inorder, IDBM_SHOW,
+		 num, _CAN_MODIFY);
+	_rec_uint8(IFACE_ERL, recs, iface, erl, IDBM_SHOW, num, _CAN_MODIFY);
+	_rec_uint32(IFACE_MAX_RECV_DLEN, recs, iface, max_recv_dlength,
+		    IDBM_SHOW, num, _CAN_MODIFY);
+	_rec_uint32(IFACE_FIRST_BURST, recs, iface, first_burst_len, IDBM_SHOW,
+		    num, _CAN_MODIFY);
+	_rec_uint16(IFACE_MAX_R2T, recs, iface, max_out_r2t, IDBM_SHOW, num,
+		    _CAN_MODIFY);
+	_rec_uint32(IFACE_MAX_BURST, recs, iface, max_burst_len, IDBM_SHOW, num,
+		    _CAN_MODIFY);
+	_rec_str(IFACE_CHAP_AUTH, recs, iface, chap_auth, IDBM_SHOW, num,
+		 _CAN_MODIFY);
+	_rec_str(IFACE_BIDI_CHAP, recs, iface, bidi_chap, IDBM_SHOW, num,
+		 _CAN_MODIFY);
+	_rec_str(IFACE_STRICT_LOGIN_COMP, recs, iface, strict_login_comp,
+		 IDBM_SHOW, num, _CAN_MODIFY);
+	_rec_str(IFACE_DISCOVERY_AUTH, recs, iface, discovery_auth, IDBM_SHOW,
+		 num, _CAN_MODIFY);
+	_rec_str(IFACE_DISCOVERY_LOGOUT, recs, iface, discovery_logout,
+		 IDBM_SHOW, num, _CAN_MODIFY);
+}
+
+static void _idbm_recs_print(struct idbm_rec *recs, FILE *f, int show)
+{
+	int i;
+	fprintf(f, "%s\n", ISCSI_BEGIN_REC);
+	for (i = 0; i < MAX_KEYS; i++) {
+		if (recs[i].visible == IDBM_HIDE)
+			continue;
+		if (!show && recs[i].visible == IDBM_MASKED) {
+			if (*(char*)recs[i].data) {
+				fprintf(f, "%s = ********\n", recs[i].name);
+				continue;
+			}
+			/* fall through */
+		}
+
+		if (strlen(recs[i].value))
+			fprintf(f, "%s = %s\n", recs[i].name, recs[i].value);
+		else if (f == stdout)
+			fprintf(f, "%s = <empty>\n", recs[i].name);
+	}
+	fprintf(f, "%s\n", ISCSI_END_REC);
+}
+
+void _idbm_iface_print(struct iscsi_iface *iface, FILE *f)
+{
+	struct idbm_rec *recs = NULL;
+
+	recs = _idbm_recs_alloc();
+	if (recs == NULL)
+		return;
+
+	_idbm_iface_rec_link(iface, recs);
+
+	_idbm_recs_print(recs, f, IDBM_SHOW);
+
+	_idbm_recs_free(recs);
+}
+
+static int _idbm_rec_update_param(struct iscsi_context *ctx,
+				  struct idbm_rec *recs, char *name,
+				  char *value, int line_number)
+{
+	int rc = LIBISCSI_OK;
+	int i = 0;
+	int passwd_done = 0;
+	char passwd_len[8];
+
+	assert(ctx != NULL);
+	assert(recs != NULL);
+	assert(name != NULL);
+	assert(value != NULL);
+
```

Comment:
Will `asserts` be enabled in released code?

Reply:
Yes. And documented as:
`If this pointer is NULL, your program will be terminated by assert.`

For internal functions, assert could help us find the error easily at early stage.